src/core/components/floodsub.js

```diff
@@ -0,0 +1,94 @@
+'use strict'
+
+const FloodSub = require('libp2p-floodsub')
+const promisify = require('promisify-es6')
+const Readable = require('stream').Readable
+
+const OFFLINE_ERROR = require('../utils').OFFLINE_ERROR
+const FSUB_ERROR = new Error(`FloodSub is not started.`)
+
+module.exports = function floodsub (self) {
+  return {
+    start: promisify((callback) => {
+      if (!self.isOnline()) {
+        throw OFFLINE_ERROR
+      }
+
+      self._floodsub = new FloodSub(self._libp2pNode)
+      return callback(null, self._floodsub)
+    }),
+
+    subscribe: promisify((topic, options, callback) => {
+      // TODO: Clarify with @diasdavid what to do with the `options.discover` param
+      // Ref: https://github.com/ipfs/js-ipfs-api/pull/377/files#diff-f0c61c06fd5dc36b6f760b7ea97b1862R50
+      if (typeof options === 'function') {
+        callback = options
+        options = {}
+      }
+
+      if (!self.isOnline()) {
+        throw OFFLINE_ERROR
+      }
+
+      if (!self._floodsub) {
+        throw FSUB_ERROR
+      }
+
+      let rs = new Readable()
+      rs.cancel = () => self._floodsub.unsubscribe(topic)
+
+      self._floodsub.on(topic, (data) => {
+        rs.emit('data', {
+          data: data.toString(),
+          topicIDs: [topic]
+        })
+      })
```

Comment:
This might cause an error that if I call subscribe on the same topic twice, I'll start seeing the same message being emitted twice as well